### PR TITLE
Migration to CE3 and other major dependencies upgrade

### PR DESCRIPTION
**Pull request by pomadchin**
# Overview

This PR migrates GT to:
- Spark 3.2.x 
- CE3
- Cassandra 4
-  JTS 1.19 

It is a breaking change, but we may release it as GT 3.7.0 🤷‍♂️ 

In CE3 we have a concept of `IORuntime`, which incapsulates threadpools for blocking, and non blocking computations. There is no need in passing a confusing `ExecutionContext` around anymore.

Closes https://github.com/locationtech/geotrellis/issues/3395 and a small step towards https://github.com/locationtech/geotrellis/issues/3388
Closes https://github.com/locationtech/geotrellis/issues/3463
Closes https://github.com/locationtech/geotrellis/issues/3382

I decided to bump all major dependencies in terms of this PR, including Cassandra and JTS.

## CQs
- [ ] GeoTools 25.4
  - See https://github.com/locationtech/geotrellis/issues/3470 for why not 27.x
- [ ] Cassandra 4.14.1
- [ ] HBase 2.4.13
- [ ] Cats Effect 3.3.13
- [ ] Cats 2.8.0
- [ ] Circe 0.14.2
- [ ] FS2 3.2.10
- [ ] Pureconfig 0.17.1
- [ ] Scala XML 2.1.0
- [ ] commons-io 2.11.0
- [ ] AWS SDK S3 V2 2.17.228
- [ ] Avro 1.10.2
- [ ] scala-parser-combinators 2.1.1
- [ ] Caffeine 2.9.3
- [ ] woodstox-core 6.3.0
- [ ] commons-configuration2 2.8.0
- [ ] re2j 1.7
- [ ] protobuf-java 3.21.2
- [ ] squants 1.8.3
- [ ] scalactic 3.2.12
- [ ] shapeless 2.3.9
- [ ] unit-api 2.1.3
- [ ] scala-uri 4.0.2
- [ ] scala-collection-compat 2.8.0
- [ ] Circe Json Schema 0.2.0
- [ ] opencsv 5.6
- [ ] Scaffeine 4.1.0

## Checklist

- [ ] CQs (merge with https://github.com/locationtech/geotrellis/pull/3452)
- [x] Dependencies updates to align all libs with the fresh cats version
- [x] [./CHANGELOG.md](https://github.com/locationtech/geotrellis/blob/master/CHANGELOG.md) updated, if necessary. Link to the issue if closed, otherwise the PR.

### CODE REVIEW

File: store/src/main/scala/geotrellis/store/util/IORuntimeTransient.scala [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+ * limitations under the License.
+ */
+
+package geotrellis.store.util
+
+import cats.effect._
+
+object IORuntimeTransient extends Serializable {
+  val ThreadsNumber: Int = Runtime.getRuntime.availableProcessors
+  @transient lazy val IORuntime: unsafe.IORuntime = unsafe.IORuntime.global
```
Comment (pomadchin) [inline store/src/main/scala/geotrellis/store/util/IORuntimeTransient.scala:23]: We still need a companion object to be sure the runtime init is a `@transient lazy val`, to ensure that it is not serialized and the instance is created per machine.

File: s3-spark/src/main/scala/geotrellis/spark/store/s3/SaveToS3.scala [hunk 2]
```diff
@@ -74,7 +72,7 @@ object SaveToS3 {
     rdd.foreachPartition { partition =>
       val s3client = s3Client
       val requests: fs2.Stream[IO, (PutObjectRequest, RequestBody)] =
-        fs2.Stream.fromIterator[IO](
+        fs2.Stream.fromBlockingIterator[IO](
```
Comment (pomadchin) [inline s3-spark/src/main/scala/geotrellis/spark/store/s3/SaveToS3.scala:75]: By the `S3AsyncClient` usage we may eliminate a lot of `blocking` calls, simplify the code, and make it more resilient, see https://github.com/locationtech/geotrellis/issues/3467

File: project/Dependencies.scala [hunk 3]
```diff
@@ -17,12 +17,12 @@
 import sbt._
 
 object Version {
-  val geotools    = "25.0"
+  val geotools    = "25.4"
```
Comment (pomadchin) [inline project/Dependencies.scala:20]: I could not upgrade it further, see: https://github.com/locationtech/geotrellis/issues/3470

File: cassandra-spark/src/main/scala/geotrellis/spark/store/cassandra/CassandraRDDReader.scala [hunk 4]
```diff
@@ -63,35 +61,35 @@ object CassandraRDDReader {
...
-            IOUtils.parJoin[K, V](seq.iterator) { index: BigInt =>
-              val row = session.execute(statement.bind(index: BigInteger))
-              if (row.asScala.nonEmpty) {
-                val bytes = row.one().getBytes("value").array()
-                val recs = AvroEncoder.fromBinary(kwWriterSchema.value.getOrElse(_recordCodec.schema), bytes)(_recordCodec)
-                if (filterIndexOnly) recs
-                else recs.filter { row => includeKey(row._1) }
-              } else Vector.empty
+            IOUtils.parJoinIO[K, V](seq.iterator) { index: BigInt =>
+              session.executeF[IO](statement.bind(index.asJava)).map { row =>
```
Comment (pomadchin) [inline cassandra-spark/src/main/scala/geotrellis/spark/store/cassandra/CassandraRDDReader.scala:80]: It's possible to implement it in a similar fashion for S3, see https://github.com/locationtech/geotrellis/issues/3467

File: project/Settings.scala [hunk 5]
```diff
@@ -133,6 +133,11 @@ object Settings {
           <name>Rob Emanuele</name>
           <url>https://github.com/lossyrob/</url>
         </developer>
+        <developer>
+          <id>pomadchin</id>
+          <name>Grigory Pomadchin</name>
```
Comment (echeipesh) [inline project/Settings.scala:138]: 💯

**Comment by pomadchin** [comment]
If someone has any time / will to help with [CQs](https://github.com/locationtech/geotrellis/blob/fa88aeddab0849cdb8f830d3747554bc7adceb1f/docs/CONTRIBUTING.rst#contribution-questionnaire) that is very much appreciated :D

**Comment by pomadchin** [comment]
I think I'm going to merge this in, and later handle CQs. It's a pretty major change so lets base all the new work on top of it.

`3.7.0` 🚀

**Comment by pomadchin** [comment]
FYI IPZilla is deprecated.

**Pull request merged**